nvaccess/nvda
# Do not report "baseline" on formatting information request

### Link to issue number:

Fixes #11815
Fixes #11078.

### Summary of the issue:

When fetching formatting information with NVDA+F (once or twice) in a Firefox page, "baseline" is reported.
Only subscript or superscript should be reported, not "baseline" since "baseline" should be considered the default text position.
Also, when "Report superscript and subscript" option is checked, navigating in a Firefox page cause spurious reporting of 'baseline'. This is due to transitions between elements having no formatting information such as button or graphics and baseline text.

### Description of how this pull request fixes the issue:

Updated design after review comments in this PR:
* Ensured that all object generating the textInfo format field provide only standard values for 'text-position' field; superscript, subscript, baseline and unspecified.
* Unspecified is used in the following cases:
  - the API does not provide formatting info (e.g. buttons in Firefox)
  - mixed text-position information, e.g. an Excel cell containing baseline and superscript.

### Alternative solution

The old design, initially implemented in this issue before review comments, was the following:
* No modification was done on the values produced by the various API.
* In `getFormatFieldSpeech`in `speech\__init__.py`, when getting 'text-position' attribute, replace any attribute that do not match 'sub' or 'super' with None.
* This allows not to report "baseline" when requesting formatting information, but also not to report "baseline" when the cursor moves from a web element that do not have any formatting attribute available.

### Testing performed:

For the folling APIs/files, the following test were made:
- Check that NVDA+F is able to report superscript and subscript
- Check that NVDA+F does not report text position information for baseline text or unspecified text position (if applicable)
- Activate "Report superscript and subscript" option and navigate whith the caret to test all possible transitions between the 3 or 4 possible text position (superscript, subscript, baseline and unspecified).

API/files tested:
* `NVDAObjects/IAccessible/__init__.py`: tested with Chrome
* `NVDAObjects/UIA/__init__.py`: tested with Word (UIA)
* `NVDAObjects/window/edit.py`, class `ITextDocumentTextInfo`: Tested with Wordpad.
* `NVDAObjects/window/edit.py`, class `EditTextInfo`: I do not know where I can test it natively, but I have used the suggestion made in https://github.com/nvaccess/nvda/pull/11897#issuecomment-932776162 to test (successfully) in WordPad.
* `NVDAObjects/window/excel.py`: tested with Excel, cell navigation.
  Note: When editing the cell's content, no formatting can be reported (neither before this PR nor with this PR).
* `NVDAObjects/window/winword.py`: tested with Word '16.0.5188.1000' (UIA not checked)
* `appModules/powerpnt.py`: tested with PowerPoint slide writing mode (formatting not reported in slideshow mode, neither before this PR nor after).
* `appModules/soffice.py`: tested with Libre Office Writer 7.1
* `virtualBuffers/MSHTML.py`: Tested with IE, with CSS `vertical-align` property.
  Note: `<sub>` / `<super>` tags do not work, neither with this PR nor before.

API/files not (yet) tested:
* `NVDAObjects/JAB/__init__.py`
In which control/program could I test?

Other tests:
* Firefox: With "Report superscript and subscript" option checked, navigate with the virtual cursor on a GitHub issue page around buttons and graphics. No spurious "baseline" can be heard.
* Excel: While editing cell content and pressing NVDA+F, "No formatting information" is now reported, since it cannot be retrieved. Before this PR, "baseline" was reported whatever the content of the cell.

### Known issues with pull request:

Two APIs have not been tested (cf Test section).
Reviewers: please help me complete the test coverage.

Also reviewer, please check that I haven't forgotten any API.

### Change log entry:

Changes
`'Baseline' is not reported anymore upon text formatting request (NVDA+F). (#11815)``

Bug fixes
`'Baseline' is not frequently spuriously reported anymore whild reading some webpages when report superscript and subscript option is active. (#11078)`
